[RFC] First draft for a contrib test suite + test for timm contrib

Wauplin opened the pull request:
First discussed in https://github.com/huggingface/huggingface_hub/issues/1190. Goal is to detect proactively breaking changes and deprecation warnings in downstream libraries.

This is a very first implementation with 1 test for `timm` library (after https://github.com/rwightman/pytorch-image-models/pull/1547) to validate the concept. I think for a start we can have a github workflow only triggered manually. It doesn't really make sense to run it on each PR or on main (will most probably always fail at least 1 integration).


For now downstream libraries are listed in a `requirements.txt` file. I'm not sure yet what's the best way to pin the versions. Maybe we could take advantage of the ["inputs" feature](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#providing-inputs) from github workflow to manually set which version we want to test. That can be done in a later PR when the use case is really needed.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1200). All of your documentation changes will be reflected on that endpoint.

osanseviero commented on contrib/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+pytest-xdist
+
+# Timm
+git+https://github.com/rwightman/pytorch-image-models.git#egg=timm
```
A challenge will be that different libraries might have conflicting versions of dependencies

osanseviero commented on .github/workflows/contrib-tests.yml:
```diff
@@ -0,0 +1,28 @@
+
+      # Install downstream libraries
+      - name: Install downstream libraries
+        run: pip install -r contrib/requirements.txt
```
For multiple versions instead of inputs we could maybe use a github actions matrix to go through different ones?

Wauplin replied:
You're right, will be a nightmare to handle. Let's have separate venvs (see https://github.com/huggingface/huggingface_hub/pull/1200#issuecomment-1322444901)

osanseviero commented on contrib/test_timm.py:
```diff
@@ -0,0 +1,10 @@
+def test_push_to_hub(repo_name: str, cleanup_repo: None) -> None:
```
I would also add a test that shows that you can download models

Wauplin replied:
Yes! Will do it once we validate the architecture (I have added a TODO in PR description for it).

osanseviero commented on contrib/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+pytest-xdist
+
+# Timm
+git+https://github.com/rwightman/pytorch-image-models.git#egg=timm
```
Timm already installs a huggingface_hub version I imagine, so how would that conflict with the pip install . from before?

Wauplin replied:
Yep, huggingface_hub has to be installed in last. Thanks for review :+1:

osanseviero reviewed (commented):
Very cool! I left some minor questions, but overall this looks to be going in the right direction! 🔥

Wauplin commented:
@osanseviero thanks for your feedback, it helped a lot ! 
I make some changes to the `contrib` structure. Now I have a common `requirements.txt`. For each lib, I have a folder with a test file and another `requirements.txt`. The workflow is now to launch 1 process/job per dependent library:
- create virtualenv for timm
- install common requirements
- install timm requirements
- uninstall huggingface_hub (if any)
- install huggingface_hub from source code
- run timm tests

This way we don't have to handle conflicts between different dependencies / versions. It makes local tests a bit more complex as it requires 1 env per contrib. Hopefully not much people will really need to run those stuff locally. What I plan to do is to add a script (either separate, either in makefile) to handle all the stuff with the venvs. WDYT in general ?

Wauplin commented:
Refactored a bit the `makefile`. Now possible to do:

```sh
# Setup all virtualenvs
make contrib_setup

# Run all tests
make contrib_tests

# Setup and run all tests at once
make contrib

# Delete all virtual envs (if corrupted)
make contrib_clear
```

And for a specific lib:

```sh
# Setup timm tests
make contrib_setup_timm

# Run timm tests
make contrib_test_timm

# Setup and run timm tests at once
make contrib_timm

# Delete timm virtualenv
make contrib_clear_timm
```

osanseviero commented on contrib/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Run contrib tests on CI
+
+Contrib tests can be [manually triggered in github](https://github.com/huggingface/huggingface_hub/actions) with the `Contrib tests` workflow.
```
```suggestion
Contrib tests can be [manually triggered in GitHub](https://github.com/huggingface/huggingface_hub/actions) with the `Contrib tests` workflow.
```

osanseviero reviewed (commented):
Looking neat! I want to make a second pass through this PR and also let's see if others have any thoughts

osanseviero commented on contrib/timm/test_timm.py:
```diff
@@ -0,0 +1,19 @@
+MODEL_ID = "nateraw/timm-resnet50-beans"
+
+
+def test_load_and_push_to_hub(repo_name: str, cleanup_repo: None) -> None:
```
Should this be 2 tests?

Wauplin replied:
Done in https://github.com/huggingface/huggingface_hub/pull/1200/commits/173aff85bd26254a9a2d0d2cbc14f0d7c1f355f0.

osanseviero commented on contrib/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@contextlib.contextmanager
+def production_endpoint() -> Generator:
```
Is this really needed? With `HF_ENDPOINT` you could change the endpoint you're using

Wauplin replied:
> Is this really needed? 

I'd say yes, kinda the same as what already exists in the `hfh tests/`. The problem with `HF_ENDPOINT` is that it is evaluated only once at startup. What I want here is to make all calls to staging environment (especially pushing to repos) except some calls that have to be made to production environment (especially loading models).

Another solution could be to upload test models to staging but then we wouldn't notice if a model changed in production.

Wauplin commented on .github/workflows/contrib-tests.yml:
```diff
@@ -0,0 +1,48 @@
+  workflow_dispatch:
+  push:
+    branches:
+      - ci_contrib_*
```
Removed in https://github.com/huggingface/huggingface_hub/pull/1200/commits/173aff85bd26254a9a2d0d2cbc14f0d7c1f355f0.

Wauplin commented on .github/workflows/contrib-tests.yml:
```diff
@@ -0,0 +1,48 @@
+  push:
+    branches:
+      - ci_contrib_*
+      - 1190-rfc-add-contrib-test-suite
```
Removed in https://github.com/huggingface/huggingface_hub/pull/1200/commits/173aff85bd26254a9a2d0d2cbc14f0d7c1f355f0.

Wauplin commented on .github/workflows/contrib-tests.yml:
```diff
@@ -0,0 +1,48 @@
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.8
```
Yeah, didn't want to do the all github matrix but definitely something we can change in case a lib doesn't support 3.8. Let's adapt later.

Wauplin commented on .github/workflows/contrib-tests.yml:
```diff
@@ -0,0 +1,48 @@
+      - name: Install common dependencies
+        run: pip install -r contrib/requirements.txt
+
+      # Install downstream library
```
Done in https://github.com/huggingface/huggingface_hub/pull/1200/commits/173aff85bd26254a9a2d0d2cbc14f0d7c1f355f0.

Wauplin commented:
Thanks for the review @osanseviero. I made some changes and addresses all of your comments.
Let's have a final review from @LysandreJik once he is back from holidays and then merge the PR.

LysandreJik commented on .github/workflows/contrib-tests.yml:
```diff
@@ -0,0 +1,47 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        contrib: [
+          "sentence_transformers",
+          "timm",
+        ]
```
Great idea to use a matrix here!

LysandreJik replied:
Should this matrix be governed by a `ls` first so that we have individual jobs for each folder under `contrib` and without the need to specify each folder here? (nitpick though, this should (or shouldn't) be done in a follow up PR)

Wauplin replied:
Added in https://github.com/huggingface/huggingface_hub/pull/1200/commits/f20ab77919533fa1254d4e89df3884956644c259 and https://github.com/huggingface/huggingface_hub/pull/1200/commits/6cb96e7658c72922dfb5cf9b7b7773e0df5fcf53 a script that list contrib tests and update the Makefile and github workflow file accordingly. Script is integrated with `make quality` and `make style` to make it easy for contributors.

LysandreJik commented on contrib/README.md:
```diff
@@ -0,0 +1,71 @@
+1. Create a subfolder with the lib name. Example: `./contrib/transformers`
+2. Create a `requirements.txt` file specific to this lib. Example `./contrib/transformers/requirements.txt`
+3. Implements tests for this lib. Example: `./contrib/transformers/test_push_to_hub.py`
+4. Edit `makefile` to add the lib to `CONTRIB_LIBS` variable. Example: `CONTRIB_LIBS := timm transformers`
+5. Edit `.github/workflows/contrib-tests.yml` to add the lib to `matrix.contrib` list. Example: `contrib: ["timm", "transformers"]`
```
I'd eventually look into automating these two so that it's slightly less error-prone, but as said above: ***nitpick***

Wauplin replied:
Now done by `make style` and `make quality`. Good call to reduce contribution efforts.

LysandreJik commented on contrib/README.md:
```diff
@@ -0,0 +1,71 @@
+
+Contrib tests can be [manually triggered in GitHub](https://github.com/huggingface/huggingface_hub/actions) with the `Contrib tests` workflow.
+
+Tests are not run in the default test suite (for each PR) as this would slow down development process. The goal is to notice breaking changes, not to avoid them. In particular, it is interesting to trigger it before a release to make sure it will not cause too much friction.
```
We could also run them once a week just to check

LysandreJik commented on contrib/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+pytest
+pytest-env
```
imo we could just require the `testing` extra here instead of adding another `requirements.txt` file.

LysandreJik commented on contrib/sentence_transformers/test_sentence_transformers.py:
```diff
@@ -0,0 +1,34 @@
+    print("Similarity:", util.dot_score(query_embedding, passage_embedding))
+
+
+@pytest.mark.xfail(
+    reason=(
+        "Production endpoint is hardcoded in sentence_transformers when pushing to Hub."
+    )
+)
+def test_push_to_hub(
+    multi_qa_model: SentenceTransformer, repo_name: str, cleanup_repo: None
+) -> None:
+    multi_qa_model.save_to_hub(repo_name)
```
Would be nice to eventually have a test that doesn't fail to ensure that save to hub actually works :) we could have a specific org for that, like skops does, but I understand it's a bit complex to setup + very annoying to have testing artifacts on the actual hub

Wauplin replied:
(discussed offline) Decision has been taken that the `contrib/` test suite purpose is only to test the deprecation warnings in downstream libraries. Testing the validity of a pushed/downloaded model is therefore out of scope here.
This can be reevaluated in the future :)

Wauplin replied:
> Would be nice to eventually have a test that doesn't fail to ensure that save to hub actually works :)

Yes, completely agree on that. I'd like to do that later. I am about to open an issue/PR on sentence transformers side to test that properly.
Worse case scenario, I have set a reminder to myself in 10 days.

LysandreJik commented on contrib/timm/test_timm.py:
```diff
@@ -0,0 +1,20 @@
+    _ = timm.create_model("hf_hub:" + MODEL_ID, pretrained=True)
+
+
+def test_push_to_hub(repo_name: str, cleanup_repo: None) -> None:
+    model = timm.create_model("resnet18")
+    timm.models.hub.push_to_hf_hub(model, repo_name)
```
Should we also test that the model pushed is according to what we expect? For example, that we can redownload it and use it once again, as this would be the usual workflow?

LysandreJik reviewed (approved):
Looks great! I left a few comments

Wauplin commented:
@osanseviero @LysandreJik thanks for the last reviews :)
I've made some last minute changes:
- added a script to list contrib test and update Makefile/github action. It's now integrated to `make style` and `make quality`.
- added a cron to run tests every Saturday midnight. Let's see how it goes. Later work should be to keep us updated about the cron result (slack message for instance ?). This can be done later as my main concern is to actually get more contrib tests first.
- removed the `contrib/requirements.txt` in favor of `.[testing]`. Let's see if at some point we need a more specific requirement.

So I think we are now finally good to go :smile: :fire:
